### PR TITLE
Implement send/recv retransmit

### PR DESCRIPTION
**Pull request by ztiffany**
-Implement send/recv retransmission
-Enhance RMA and send/recv criterion tests
  -Check counters, full cqe data, err cqes
-Implement send/recv selective completion

Signed-off-by: Zach Tiffany ztiffany@cray.com

@hppritcha @sungeunchoi @jswaro @jshimek

### CODE REVIEW

File: prov/gni/src/gnix_msg.c [hunk 1]
```diff
@@ -120,22 +133,47 @@ static inline int __gnix_recv_completion(struct gnix_fid_ep *ep,
...
+	int rc;
+
+	flags |= req->msg.send_flags & FI_TAGGED;
+
+	if (ep->send_cq) {
+		rc = _gnix_cq_add_error(ep->send_cq, req->user_context,
+					flags, 0, 0, 0, 0, 0, FI_ECANCELED,
+					GNI_RC_TRANSACTION_ERROR, NULL);
+		if (rc != FI_SUCCESS)  {
+			GNIX_WARN((FI_LOG_CQ | FI_LOG_EP_DATA),
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_msg.c]: This is going to lead to an out of bounds array access.  The first parameter is an index (enum) into an array of strings used to print the error message, so these are not flags that can be OR'ed.

File: prov/gni/src/gnix_msg.c [hunk 2]
```diff
@@ -120,22 +133,47 @@ static inline int __gnix_recv_completion(struct gnix_fid_ep *ep,
...
+				   "_gnix_cq_add_error() returned %d\n",
```
Comment (hppritcha) [inline prov/gni/src/gnix_msg.c:149]: Sung and I don't think it works to OR together subsystems for the GNIX_WARN macro

**Comment by hppritcha** [comment]
:+1: except GNIX_WARN problem

File: prov/gni/src/gnix_msg.c [hunk 3]
```diff
@@ -120,22 +133,47 @@ static inline int __gnix_recv_completion(struct gnix_fid_ep *ep,
...
 				req->user_context,
-				FI_SEND | FI_MSG,
-				req->msg.send_len,
-				(void *)req->msg.send_addr,
-				req->msg.imm,
-				req->msg.tag);
+				flags, 0, 0, 0, 0);
 		if (rc != FI_SUCCESS)  {
 			GNIX_WARN((FI_LOG_CQ | FI_LOG_EP_DATA),
 					"_gnix_cq_add_event returned %d\n",
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_msg.c:179]: Another case of logging subsystem specification.

**Pull request merged**